### PR TITLE
[SPARK-9372] [SQL] For joins, insert IS NOT NULL filters to children.

### PR DESCRIPTION
**Pull request by nongli**
Some join types and conditions imply that the join keys cannot be NULL and
can be filtered out by the children. This patch does this for inner joins
and introduces a mechanism to generate predicates. The complex part of doing
this is to make sure the transformation is stable. The problem that we want
to avoid is generating a filter in the join, having that pushed down and then
having the join regenerate the filter.

This patch solves this by having the join remember predicates that it has
generated. This mechanism should be general enough that we can infer other
predicates, for example "a join b where a.id = b.id AND a.id = 10" could
also use this mechanism to generate the predicate "b.id = 10".

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 1]
```diff
@@ -190,14 +190,15 @@ trait CheckAnalysis {
...
-          case j @ Join(left, right, _, _) if left.outputSet.intersect(right.outputSet).nonEmpty =>
-            val conflictingAttributes = left.outputSet.intersect(right.outputSet)
-            failAnalysis(
-              s"""
-                 |Failure when resolving conflicting references in Join:
-                 |$plan
-                 |Conflicting attributes: ${conflictingAttributes.mkString(",")}
-                 |""".stripMargin)
+          case j @ Join(left, right, _, _, _)
+            if left.outputSet.intersect(right.outputSet).nonEmpty =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: nit: we can use `if !j.selfJoinResolved`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -356,6 +360,51 @@ object LikeSimplification extends Rule[LogicalPlan] {
...
+ * This rule adds IsNotNull predicates based on join keys. If the join contains a condition
+ * `a` binaryOp *, a is non-nullable. This adds filters for those attributes to the children.
+ *
+ * To avoid the problem of repeatedly generating the IsNotNull predicates, the join operator
+ * remembers all the expressions it generated.
+ */
+object AddJoinKeyNullabilityFilters extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case j @ Join(left, right, Inner, Some(condition), generated) => {
+      val nonNullableKeys = ExtractNonNullableAttributes.unapply(condition)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: nit: `val ExtractNonNullableAttributes(nonNullableKeys) = condition`

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: oh nvm, I think your version looks more clear.

**Comment by cloud-fan** [comment]
This looks very cool!
But I'm a little worried about making the `Join` stateful, how about adding a new batch in `Optimizer` and only execute it `Once`? So that we can avoid having the join regenerate the filter as your rule will be applied after all other optimization rules.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala [hunk 3]
```diff
@@ -122,11 +122,22 @@ case class Except(left: LogicalPlan, right: LogicalPlan) extends SetOperation(le
...
+  }
+}
+
 case class Join(
   left: LogicalPlan,
   right: LogicalPlan,
   joinType: JoinType,
-  condition: Option[Expression]) extends BinaryNode {
+  condition: Option[Expression],
+  generatedExpressions: Option[EquivalentExpressions]) extends BinaryNode {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: This is semi-public API cause I think some advanced projects do dig into catalyst and we've never changed the signature of something as basic as `Join` before.  Could we do this instead by fixing nullablity propagation and only inserting the filter if the attribute is `nullable`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala [hunk 4]
```diff
@@ -43,7 +43,7 @@ abstract class PlanTest extends SparkFunSuite {
   protected def comparePlans(plan1: LogicalPlan, plan2: LogicalPlan) {
     val normalized1 = normalizeExprIds(plan1)
     val normalized2 = normalizeExprIds(plan2)
-    if (normalized1 != normalized2) {
+    if (!normalized1.semanticEquals(normalized2)) {
```
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala]: Existing: do we need this hacky normalization logic above anymore?  I don't think `semanticEquals` existed when I wrote this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 5]
```diff
@@ -99,6 +99,13 @@ abstract class LogicalPlan extends QueryPlan[LogicalPlan] with Logging {
    */
   lazy val resolved: Boolean = expressions.forall(_.resolved) && childrenResolved
 
+  /**
+   * Returns true if the two plans are semantically equal. This should ignore state generated
+   * during planning to help the planning process.
+   * TODO: implement this as a pass that canonicalizes the plan tree instead?
+   */
+  def semanticEquals(other: LogicalPlan): Boolean = this == other
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: Oh, this is a new semantic equals.  How is this different than `sameResult`?  Maybe we should unify the naming between Expression and LogicalPlan for this concept.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala (hunk 3 above)
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: I started down that path but can you think of a way to make that handle the more general case of predicate propagation?

t1.key join t2.key where t1.key = t2.key and t1.key = 5.

How do we generate the predicate t2.key = 5? how do we make this more general?

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: In an earlier version of catalyst we also had equivalence classes propagate up the logical plans.  Would that give you enough information?

Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: Equivalence classes is one thing, we can compute that no problem I think. The issue is how to remember that t2.key = 5 was generated and not to generate it again. The trick of setting nullable doesn't work here. We could maintain value constraints (where nullability is a subset).

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: Couldn't `Literal(5)` be in the equivalence class and we could check for that?

That said, I also like the idea more general value constraints.

Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: Literal(5) would work for equivalence but we want to track more than equality. If it was t1.key join t2.key where t1.key = t2.key and t1.key > 5, we'd similarly want to add t2.key > 5. 

Are you suggesting we don't change the operator and walk the tree bottom up to collect these constraints? This seems extremely expensive to do.

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: I was thinking operators would propagate the set of constraints up from their children (possibly augmenting or clearing as appropriate) and we'd save it in a `lazy val`.

**Comment by srowen** [comment]
Is this already fixed by https://github.com/apache/spark/pull/7768 ?

**Pull request closed without merging**